compiler/rustc_passes/src/abi_test.rs

```diff
@@ -0,0 +1,92 @@
+use rustc_ast::Attribute;
+use rustc_hir::def::DefKind;
+use rustc_hir::def_id::DefId;
+use rustc_middle::ty::layout::{FnAbiError, LayoutError};
+use rustc_middle::ty::{self, GenericArgs, Instance, TyCtxt};
+use rustc_span::source_map::Spanned;
+use rustc_span::symbol::sym;
+
+use crate::errors::{AbiOf, UnrecognizedField};
+
+pub fn test_abi(tcx: TyCtxt<'_>) {
+    if tcx.features().rustc_attrs {
+        // if the `rustc_attrs` feature is not enabled, don't bother testing ABI
+        for id in tcx.hir().items() {
+            match tcx.def_kind(id.owner_id) {
+                DefKind::Fn => {
+                    for attr in tcx.get_attrs(id.owner_id, sym::rustc_abi) {
+                        dump_abi_of(tcx, id.owner_id.def_id.into(), attr);
+                    }
+                }
+                DefKind::Impl { .. } => {
+                    // To find associated functions we need to go into the child items here.
+                    for &id in tcx.associated_item_def_ids(id.owner_id) {
+                        if matches!(tcx.def_kind(id), DefKind::AssocFn) {
+                            for attr in tcx.get_attrs(id, sym::rustc_abi) {
+                                dump_abi_of(tcx, id, attr);
+                            }
+                        }
+                    }
+                }
+                _ => {}
+            }
+        }
+    }
+}
+
+fn dump_abi_of(tcx: TyCtxt<'_>, item_def_id: DefId, attr: &Attribute) {
+    let param_env = tcx.param_env(item_def_id);
+    let args = GenericArgs::identity_for_item(tcx, item_def_id);
+    let instance = match Instance::resolve(tcx, param_env, item_def_id, args) {
+        Ok(Some(instance)) => instance,
+        Ok(None) => {
+            // Not sure what to do here, but `LayoutError::Unknown` seems reasonable?
+            let ty = tcx.type_of(item_def_id).instantiate_identity();
+            tcx.sess.emit_fatal(Spanned {
+                node: LayoutError::Unknown(ty).into_diagnostic(),
+
+                span: tcx.def_span(item_def_id),
+            });
+        }
+        Err(_guaranteed) => return,
+    };
```

Comment:
This should always resolve to `Instance::mono(item_def_id, args)` -- not sure if you need to actually resolve?

Reply:
Then it will `bug!` on a generic item though, won't it?

Reply:
Ah, yeah, I guess it's designed to prevent you from side-stepping resolution like this.

Reply:
The neat thing is that if the generics do not affect the ABI, `rustc_abi(debug)` actually works on generic items. :)  I just added a test for that.

Reply:
Yeah, which is why I had suggested mono earlier, I thought it just constructed an InstanceDef::Item..